### PR TITLE
[IMP] mail: recent emoji in emoji picker

### PR DESCRIPTION
**Pull request by alexkuhn**
Task-2982627

Screenshot 2023-03-20 at 18 36 07

### CODE REVIEW

File: addons/mail/static/src/core/thread_service.js [hunk 1]
```diff
@@ -705,9 +707,23 @@ export class ThreadService {
...
+            );
+            const regex =
+                /(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\u303d|\u3030|\u24c2|\ud83c[\udd70-\udd71]|\ud83c[\udd7e-\udd7f]|\ud83c\udd8e|\ud83c[\udd91-\udd9a]|\ud83c[\udde6-\uddff]|\ud83c[\ude01-\ude02]|\ud83c\ude1a|\ud83c\ude2f|\ud83c[\ude32-\ude3a]|\ud83c[\ude50-\ude51]|\u203c|\u2049|[\u25aa-\u25ab]|\u25b6|\u25c0|[\u25fb-\u25fe]|\u00a9|\u00ae|\u2122|\u2139|\ud83c\udc04|[\u2600-\u26FF]|\u2b05|\u2b06|\u2b07|\u2b1b|\u2b1c|\u2b50|\u2b55|\u231a|\u231b|\u2328|\u23cf|[\u23e9-\u23f3]|[\u23f8-\u23fa]|\ud83c\udccf|\u2934|\u2935|[\u2190-\u21ff])/g;
+            const emojisInContent = prettyContent.match(regex) ?? [];
+            for (const codepoints of emojisInContent) {
+                if (emojis.some((emoji) => emoji.codepoints === codepoints)) {
+                    recentEmojis[codepoints] = (recentEmojis[codepoints] ?? 0) + 1;
+                }
+            }
+            browser.localStorage.setItem("mail.emoji.frequent", JSON.stringify(recentEmojis));
```
Comment (phenix-factory) [inline addons/mail/static/src/core/thread_service.js:723]: I feel like this is a really heavy solution that: 
* is Regex based, witch is heavy to maintain.
* Slow down the send message process, the `await loadingEmoji()` isn't the fastest function.

An alternative would be not to count send emojis, but clicked emojis from the emoji picker. That way you don't have to parse a string.

**Comment by phenix-factory** [comment]
Shouldn't the recent category be limited in size? Or older entries should be taken out after some times not used?

File: addons/mail/static/src/core/thread_service.js (hunk 1 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/thread_service.js:723]: > An alternative would be not to count send emojis, but clicked emojis from the emoji picker. That way you don't have to parse a string.

We should take emoticon (e.g. `:)`) and shortcode (e.g. `:wink:`) substitutions.

> Slow down the send message process, the await loadingEmoji() isn't the fastest function.

`loadEmoji` is just slow the first time it's called. Then it's memoized, so it's not slow at all.

**Comment by alexkuhn** [comment]
> Shouldn't the recent category be limited in size? Or older entries should be taken out after some times not used?

The display is limited to 42. We should store data on all emojis, including 43rd, 44th, etc. Otherwise you cannot correctly promote new entries.

**Comment by alexkuhn** [comment]
Moved here: https://github.com/odoo/odoo/pull/116067 (target master)

**Pull request closed without merging**